// @pushprotocol/restapi/src/lib/pushapi/profile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Profile = void 0;
const tslib_1 = require("tslib");
const PUSH_USER = require("../user");
class Profile {
    constructor(account, decryptedPgpPvtKey, env, progressHook) {
        this.account = account;
        this.decryptedPgpPvtKey = decryptedPgpPvtKey;
        this.env = env;
        this.progressHook = progressHook;
    }
    info() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const response = yield PUSH_USER.get({
                account: this.account,
                env: this.env,
            });
            return response.profile;
        });
    }
    update(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { name, desc, picture } = options;
            const response = yield PUSH_USER.profile.update({
                pgpPrivateKey: this.decryptedPgpPvtKey,
                account: this.account,
                profile: {
                    name: name,
                    desc: desc,
                    picture: picture,
                },
                env: this.env,
                progressHook: this.progressHook,
            });
            return response.profile;
        });
    }
}
exports.Profile = Profile;
