// @pushprotocol/restapi/src/lib/pushstream/pushStreamTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NOTIFICATION = exports.ProposedEventNames = exports.GroupEventType = exports.MessageEventType = exports.MessageOrigin = exports.NotificationEventType = exports.STREAM = void 0;
var STREAM;
(function (STREAM) {
    STREAM["PROFILE"] = "STREAM.PROFILE";
    STREAM["ENCRYPTION"] = "STREAM.ENCRYPTION";
    STREAM["NOTIF"] = "STREAM.NOTIF";
    STREAM["NOTIF_OPS"] = "STREAM.NOTIF_OPS";
    STREAM["CHAT"] = "STREAM.CHAT";
    STREAM["CHAT_OPS"] = "STREAM.CHAT_OPS";
    STREAM["CONNECT"] = "STREAM.CONNECT";
    STREAM["DISCONNECT"] = "STREAM.DISCONNECT";
})(STREAM = exports.STREAM || (exports.STREAM = {}));
var NotificationEventType;
(function (NotificationEventType) {
    NotificationEventType["INBOX"] = "notification.inbox";
    NotificationEventType["SPAM"] = "notification.spam";
})(NotificationEventType = exports.NotificationEventType || (exports.NotificationEventType = {}));
var MessageOrigin;
(function (MessageOrigin) {
    MessageOrigin["Other"] = "other";
    MessageOrigin["Self"] = "self";
})(MessageOrigin = exports.MessageOrigin || (exports.MessageOrigin = {}));
var MessageEventType;
(function (MessageEventType) {
    MessageEventType["Message"] = "message";
    MessageEventType["Request"] = "request";
    MessageEventType["Accept"] = "accept";
    MessageEventType["Reject"] = "reject";
})(MessageEventType = exports.MessageEventType || (exports.MessageEventType = {}));
var GroupEventType;
(function (GroupEventType) {
    GroupEventType["CreateGroup"] = "createGroup";
    GroupEventType["UpdateGroup"] = "updateGroup";
    GroupEventType["JoinGroup"] = "joinGroup";
    GroupEventType["LeaveGroup"] = "leaveGroup";
    GroupEventType["Remove"] = "remove";
})(GroupEventType = exports.GroupEventType || (exports.GroupEventType = {}));
var ProposedEventNames;
(function (ProposedEventNames) {
    ProposedEventNames["Message"] = "chat.message";
    ProposedEventNames["Request"] = "chat.request";
    ProposedEventNames["Accept"] = "chat.accept";
    ProposedEventNames["Reject"] = "chat.reject";
    ProposedEventNames["LeaveGroup"] = "chat.group.participant.leave";
    ProposedEventNames["JoinGroup"] = "chat.group.participant.join";
    ProposedEventNames["CreateGroup"] = "chat.group.create";
    ProposedEventNames["UpdateGroup"] = "chat.group.update";
    ProposedEventNames["Remove"] = "chat.group.participant.remove";
})(ProposedEventNames = exports.ProposedEventNames || (exports.ProposedEventNames = {}));
exports.NOTIFICATION = {
    TYPE: {
        BROADCAST: 1,
        TARGETTED: 3,
        SUBSET: 4,
    },
};
